src/core/reference/include/openvino/reference/paged_attention.hpp

```diff
@@ -0,0 +1,330 @@
+
+template <typename T>
+void paged_attention(
+    T* out,                               // output: attention result
+    T* score,                             // output: concatenated raw scores
+    const T* query,                       // shape: [batch_tokens, num_heads * head_size]
+    const T* key,                         // shape: [batch_tokens, num_kv_heads * head_size]
+    const T* value,                       // shape: [batch_tokens, num_kv_heads * head_size]
+    const T* key_cache,                   // shape: [num_blocks, num_kv_heads, block_size, head_size]
+    const T* value_cache,                 // shape: [num_blocks, num_kv_heads, block_size, head_size]
+    const Shape& q_shape,                 // e.g. {batch_tokens, num_heads * head_size}
+    const Shape& kv_shape,                // e.g. {batch_tokens, num_kv_heads * head_size}
+    const Shape& kv_cache_shape,          // e.g. {num_blocks, num_kv_heads, block_size, head_size}
+    const int32_t* past_lens,             // [batch_seq]: past tokens per sequence
+    const Shape& past_lens_shape,         // e.g. {batch_seq}
+    const int32_t* subsequence_begins,    // [batch_seq + 1]: start indices of new tokens per sequence
+    const int32_t* block_indices,         // [num_blocks]: block table for each sequence
+    const int32_t* block_indices_begins,  // [batch_seq + 1]: indices into block_indices per sequence
+    const T* scale_ptr,                   // (Optional) attention scale factor (can be nullptr; default = 1)
+    const int32_t* sliding_window_ptr,    // (Optional) sliding window parameter (can be nullptr; default = 0)
+    const T* alibi_slopes,  // (Optional) [num_kv_heads]: per-head bias slopes (can be nullptr; default = 0)
+    const int32_t* max_context_len_ptr,  // max context length (for score output indexing)
+
+    // Rotation parameters (if any is nullptr, rotation is skipped)
+    const int32_t* rotated_block_indices,  // (Optional) [num_rotated_blocks]: blocks to which RoPE is applied
+    const int32_t* rotation_deltas,  // (Optional) [num_rotated_blocks, block_size || 1]: indices into the trig LUT
+    const T* rotation_trig_lut,      // (Optional) LUT: [lut_rows, head_size] (first half: cosines, second half: sines)
+    const Shape& rotated_block_indices_shape,  // shape of rotated_block_indices (e.g. {num_rotated_blocks})
+    const Shape& rotation_deltas_shape,        // shape of rotation_deltas (e.g. {num_rotated_blocks, block_size} or
+                                               // {num_rotated_blocks, 1})
+    const Shape& rotation_trig_lut_shape) {    // shape of rotation_trig_lut
+
+    // Use default values for optional parameters if pointers are null.
+    T scale = scale_ptr ? scale_ptr[0] : T(1);
+    int32_t sliding_window = sliding_window_ptr ? sliding_window_ptr[0] : 0;
+    int32_t max_context_len = max_context_len_ptr ? max_context_len_ptr[0] : 0;
+
+    // Determine dimensions.
+    int32_t num_blocks = kv_cache_shape[0];
```

Comment:
```suggestion
    const auto& num_blocks = kv_cache_shape[0];
```
Why use int32 as shape has size_t type. It cannot be negative and size_t values can overflow i32.

Reply:
Fixed